packages/doenetml-worker-rust/doenetml-core/src/state_var_calculations.rs

```diff
@@ -0,0 +1,762 @@
+
+                            match new_current_freshness {
+                                // No need to recurse if the state var of the dependency is already fresh
+                                // so don't do anything to just continue the inner loop
+                                Freshness::Fresh => (),
+                                Freshness::Stale => {
+                                    // Recurse by putting two items on the stack
+                                    // and then continuing at start of stack loop
+                                    freshen_stack.push(StateVarCalculationState::Stale(
+                                        StaleCalculationState {
+                                            component_state,
+                                            instruction_idx,
+                                            val_idx,
+                                        },
+                                    ));
+                                    freshen_stack.push(StateVarCalculationState::Stale(
+                                        StaleCalculationState {
+                                            component_state: new_component_state,
+                                            instruction_idx: 0,
+                                            val_idx: 0,
+                                        },
+                                    ));
+
+                                    // break out of the inner loop and continue to outer loop,
+                                    // which will result in first popping off the last instruction
+                                    // that we added to the stack
+                                    continue 'stack_loop;
+                                }
+                                Freshness::Unresolved => {
+                                    panic!("How did a stale state variable depend on an unresolved state variable?")
+                                }
+                            };
+                        }
+                    }
+
+                    initial_val_idx = 0;
+                }
+
+                // If we have gotten to this part,
+                // then we have freshened all dependencies
+                // for the state variable (component_idx, state_var_idx).
+                // We calculate its value if a dependency has changed,
+                // else restore its previous value.
+
+                let component_idx = component_state.component_idx;
+                let state_var_idx = component_state.state_var_idx;
+
+                let mut comp = components[component_idx].borrow_mut();
+                let state_var = &mut comp.get_state_variables_mut()[state_var_idx];
+
+                // Check if any dependency has changed since we last called record_all_dependencies_viewed.
+                if state_var.check_if_any_dependency_changed_since_last_viewed() {
```

Comment:
This seems to be some magic that we haven't talked about yet?

Reply:
Each state variable has a counter that is incremented on `.set_value()`. The call to `state_var.record_all_dependencies_viewed()` records the values of the counters of the dependency state variables (including essential data). The call to `state_var.check_if_any_dependency_changed_since_last_viewed()` simply checks to see if any of the state variable counters has been incremented since the last call to `state_var.record_all_dependencies_viewed()`.

Reply:
Can these counters roll over? Or are they u64s where that practically shouldn't be an issue...?

Reply:
I used u32s. I figure 4 billion updates on a single state variable would be unlikely. But, I could change the > comparison to a != comparison just in case it rolls over.